util/topgen/templates/chiplevel.sv.tpl

```diff
@@ -0,0 +1,1198 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+${gencmd}
+<%
+import re
+import topgen.lib as lib
+from copy import deepcopy
+
+# Provide shortcuts for some commonly used variables
+pinmux = top['pinmux']
+pinout = top['pinout']
+
+num_mio_inputs = pinmux['io_counts']['muxed']['inouts'] + \
+                 pinmux['io_counts']['muxed']['inputs']
+num_mio_outputs = pinmux['io_counts']['muxed']['inouts'] + \
+                  pinmux['io_counts']['muxed']['outputs']
+num_mio_pads = pinmux['io_counts']['muxed']['pads']
+
+num_dio_inputs = pinmux['io_counts']['dedicated']['inouts'] + \
+                 pinmux['io_counts']['dedicated']['inputs']
+num_dio_outputs = pinmux['io_counts']['dedicated']['inouts'] + \
+                  pinmux['io_counts']['dedicated']['outputs']
+num_dio_total = pinmux['io_counts']['dedicated']['inouts'] + \
+                pinmux['io_counts']['dedicated']['inputs'] + \
+                pinmux['io_counts']['dedicated']['outputs']
+
+def get_dio_sig(pinmux: {}, pad: {}):
+  '''Get DIO signal associated with this pad or return None'''
+  for sig in pinmux["ios"]:
+    if pad["name"] == sig["pad"]:
+      return sig
+  else:
+    return None
+
+# Modify the pad lists on the fly, based on target config
+maxwidth = 0
+muxed_pads = []
+dedicated_pads = []
+k = 0
+for pad in pinout["pads"]:
+  if pad["muxed"]:
+    if pad["name"] not in target["stub_pads"]:
+      maxwidth = max(maxwidth, len(pad["name"]))
+      muxed_pads.append(pad)
+  else:
+    k = pad["idx"]
+    if pad["name"] not in target["stub_pads"]:
+      maxwidth = max(maxwidth, len(pad["name"]))
+      dedicated_pads.append(pad)
+
+for pad in target["add_pads"]:
+  # Since these additional pads have not been elaborated in the merge phase,
+  # we need to add their global index here.
+  amended_pad = deepcopy(pad)
+  amended_pad.update({"idx" : k})
+  dedicated_pads.append(pad)
+  k += 1
+
+num_im = sum([x["width"] if "width" in x else 1 for x in top["inter_signal"]["external"]])
+
+max_sigwidth = max([x["width"] if "width" in x else 1 for x in top["pinmux"]["ios"]])
+max_sigwidth = len("{}".format(max_sigwidth))
+
+clks_attr = top['clocks']
+cpu_clk = top['clocks']['hier_paths']['top'] + "clk_proc_main"
+cpu_rst = top["reset_paths"]["sys"]
+dm_rst = top["reset_paths"]["lc"]
+esc_clk = top['clocks']['hier_paths']['top'] + "clk_io_div4_timers"
+esc_rst = top["reset_paths"]["sys_io_div4"]
+
+unused_resets = lib.get_unused_resets(top)
+unused_im_defs, undriven_im_defs = lib.get_dangling_im_def(top["inter_signal"]["definitions"])
```

Comment:
Can we put all this Python code into topgen and pass it to the template? There is already a massive amount of magic going on, let's not add more.

(Ultimately, I want to get rid of passing the whole opaque top struct to the template, creating a very tight and hard to understand coupling between the top.hjson and the template -- but that's for another day.)

Reply:
Yeah I know there is a lot of magic going on here in these templates.

I've already tried to encapsulate larger python segments in functions in order to shove them out into `lib` or `topgen`, but I agree that more could be done.

I'll do another cleanup pass once all CI checks pass again.